### PR TITLE
fix incorrect memory limit and usage metric in container enviroment

### LINKED ISSUE #31262: MemTotal inference is flawed when container cgroup memory limit is not set
## Bug Report

When TiDB runs in container, it will read `/sys/fs/cgroup/memory/memory.limit_in_bytes` to infer memory limit. However, when the cgroup memory limit is not set, the inferred MemTotal would be `9223372036854771712` (LONG_MAX, rounded with page size), which makes the [memory alarm](https://github.com/pingcap/tidb/blob/38355172ee7235ffda612ddadf99e47d60b3b051/util/expensivequery/memory_usage_alarm.go#L115) invalid.

According to the code, this inference is also used by `PreparedPlanCacheMaxMemory`, so I suspect that limit would also suffers in container when memory limit is not set.

### 1. Minimal reproduce step (Required)
1. Run TiDB in container
2. Push TiDB memory usage to the MemTotal of `host`
3. TiDB memory usage exceed 80% but no heap dump occurs according to the log

### 2. What did you expect to see? (Required)
When the memory limit of container is not set, the memory alarm should fallback to host's MemTotal and trigger heap dump when the memory usage / host MemTotal reaches alarm ratio.

### 3. What did you see instead (Required)
When the memory limit of container is not set, the memory alarm cannot be properly triggered.

### 4. What is your TiDB version? (Required)
`v5.3.0`

### PR DESCRIPTION
**Pull request by bo-er**
### What problem does this PR solve?

1. TIDB now supports collection memory info for cgroupv2

2. the memory limit is always the maximum value of the Linux virtual memory system, which is 2^63 - 4096
   when memory is not limited by cgroup, now it's fixed by using the host's total memory.

3. using "hierarchical_memory_limit" from "memory.stat", when memory.use_hierarchy is set to 1 (otherwise the memory limit value is incorrect,
   and very likely it's gonna be the maximum value 2^63 - 4096)

Issue Number: close #31262

Problem Summary:

1. TIDB only supports collecting memory info for cgroupv1
2. the memory limit is always the maximum value of the Linux virtual memory system, which is 2^63 - 4096
   when memory is not limited by cgroup
3. the memory limit value is incorrect if simply read from memory.limit_in_bytes when memory.use_hierarchy is set to 1

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note

Fix the issue that the memory alarm would be invalid under container deployment of TIDB when the container is not being restricted by cgroup, as a result of not having correctly collected memory-related metrics.

```

### CODE REVIEW

File: util/memory/meminfo.go [hunk 1]
```diff
@@ -200,3 +185,232 @@ func InstanceMemUsed() (uint64, error) {
...
+			return maximumBytes, err
+		}
+		if limitInBytes == uint64(maximumBytes) {
+			if isHierarchical {
+				return grepCgroupFile(memCtrName, "memory.stat", "hierarchical_memory_limit")
+			}
+		}
+		return limitInBytes, nil
+	}
+	// for cgroup v2, we get memory limit directly from /sys/fs/cgroup/memory.max
```
Comment (wshwsh12) [inline util/memory/meminfo.go:234]: What will happen if tidb process is attached to a child cgroup?
For example, /sys/fs/cgroup/xx/yy/zz/memory.max. We also use the value of `/sys/fs/cgroup/memory.max`?

File: util/memory/meminfo.go [hunk 2]
```diff
@@ -200,3 +185,232 @@ func InstanceMemUsed() (uint64, error) {
...
+		return "", err
+	}
+	if isCgroupV1 {
+		gf = func(line string) bool {
+			return strings.Contains(line, controller)
+		}
+	} else { // cgroup v2
+		gf = func(line string) bool {
+			fields := strings.Split(line, ":")
+			return len(fields) == 3 && fields[0] == "0" && fields[1] == ""
```
Comment (wshwsh12) [inline util/memory/meminfo.go:316]: Please add some comments

**Review by wshwsh12 — COMMENTED** [review]
Please add some test case to check the result of parsing `/proc/self/cgroup` is correct.

**Comment by chaosue** [comment]
i have the same trouble on cgroupv2. when could this PR be merged 😊 ? 

`[2022/11/15 17:35:32.049 +08:00] [FATAL] [terror.go:292] ["unexpected error"] [error="open /sys/fs/cgroup/memory/memory.limit_in_bytes: no such file or directory"] [stack="github.com/pingcap/tidb/parser/terror.MustNil\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/parser/terror/terror.go:292\ngithub.com/pingcap/tidb/util/memory.init.0\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/memory/meminfo.go:144\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6222\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:233"] [stack="github.com/pingcap/tidb/parser/terror.MustNil\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/parser/terror/terror.go:292\ngithub.com/pingcap/tidb/util/memory.init.0\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/util/memory/meminfo.go:144\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6222\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:6199\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:233"`

File: util/memory/meminfo.go [hunk 3]
```diff
@@ -200,3 +185,232 @@ func InstanceMemUsed() (uint64, error) {
...
+// find more information on https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt
+// cgroup v1 only
+func getHierarchicalStatus() bool {
+	value, _ := readCgroupFile(memCtrName, "memory.use_hierarchy")
+	return strings.TrimSpace(string(value)) == "1"
+}
+
+// isCgrpV1 tells whether we are using cgroupV1
+// see https://docs.docker.com/config/containers/runmetrics/#enumerate-cgroups
+func isCgrpV1() (bool, error) {
```
Comment (luodw) [inline util/memory/meminfo.go:282]: here can also refer to runc, how to judge whether  in cgroupv2 https://github.com/opencontainers/runc/blob/main/libcontainer/cgroups/utils.go#L34

**Pull request closed without merging**